draft-ietf-tls-esni.md

```diff
@@ -572,12 +574,57 @@ trigger retries, as described in {{handle-server-response}}. This may be
 implemented, for instance, by reporting a failed connection with a dedicated
 error code.
 
-### GREASE extensions {#grease-extensions}
+### HelloRetryRequest
+
+If the server sends a HelloRetryRequest in response to the ClientHello,
+the client sends a second updated ClientHello per the rules in {{RFC8446}}.
+However, at this point, the client does not know whether the server processed
+ClientHelloOuter or ClientHelloInner, and MUST regenerate both values to
+be acceptable. Note: if the inner and outer ClientHellos use different groups
+for their key shares or differ in some other way, then the HelloRetryRequest may
+actually be invalid for one or the other ClientHello. If the inner ClientHello
+is unaffected by this retry, then the client only changes the outer ClientHello.
+In contrast, if the outer ClientHello is unaffected by this retry, then the client
+changes the inner ClientHello and recomputes any fields necessary in the outer
```

Comment:
This seems like an unwise optimization. What's the rationale here?

Also, won't it be impossible because you'll get a different echo_nonce?

Reply:
By optimization do you mean it'd be simpler to just re-generate both? I think the intent was to stick with the HRR model and change as little as possible. But I'd be fine with regenerating both inner and outer CHs. That would definitely be simpler to implementer, IMO.